src/Infrastructure.EntityFramework/Repositories/CollectionRepository.cs

```diff
@@ -473,6 +473,97 @@ public async Task DeleteManyAsync(IEnumerable<Guid> collectionIds)
         }
     }
 
+    public async Task CreateOrUpdateAccessForManyAsync(Guid organizationId, IEnumerable<Guid> collectionIds,
+        IEnumerable<CollectionAccessSelection> users, IEnumerable<CollectionAccessSelection> groups)
+    {
+        using (var scope = ServiceScopeFactory.CreateScope())
+        {
+            var dbContext = GetDatabaseContext(scope);
+
+            var collectionIdsList = collectionIds.ToList();
+
+            if (users != null)
+            {
+                var existingCollectionUsers = await dbContext.CollectionUsers
+                    .Where(cu => collectionIdsList.Contains(cu.CollectionId))
+                    .ToDictionaryAsync(x => (x.CollectionId, x.OrganizationUserId));
+
+                var requestedUsers = users.ToList();
+
+                foreach (var collectionId in collectionIdsList)
+                {
+                    foreach (var requestedUser in requestedUsers)
+                    {
+                        if (!existingCollectionUsers.TryGetValue(
+                                (collectionId, requestedUser.Id),
+                                out var existingCollectionUser)
+                            )
+                        {
+                            // This is a brand new entry
+                            dbContext.CollectionUsers.Add(new CollectionUser
+                            {
+                                CollectionId = collectionId,
+                                OrganizationUserId = requestedUser.Id,
+                                HidePasswords = requestedUser.HidePasswords,
+                                ReadOnly = requestedUser.ReadOnly,
+                                Manage = requestedUser.Manage
+                            });
+                            continue;
+                        }
+
+                        // It already exists, update it
+                        existingCollectionUser.HidePasswords = requestedUser.HidePasswords;
+                        existingCollectionUser.ReadOnly = requestedUser.ReadOnly;
+                        existingCollectionUser.Manage = requestedUser.Manage;
+                        dbContext.CollectionUsers.Update(existingCollectionUser);
```

Comment:
Any idea on performance here, calling `Update` for each entity that we're updating, vs. adding them to a list and calling `UpdateRange` at the end? Compare this to other bulk methods in this repository, e.g. `ReplaceCollectionGroupsAsync`, which follow this pattern. I don't know for sure, but I'm inclined to follow the other methods as examples.

Same for calling `Add` vs. `AddRange`.

Reply:
Great question! I did a bit digging and it looks like there was a performance difference in EF6. However, in EF Core the `AddRange` and `UpdateRange` are [purely convenience methods and do not have a significant performance difference](https://learn.microsoft.com/en-us/ef/core/change-tracking/miscellaneous#addrange-updaterange-attachrange-and-removerange).

I'm not familiar with exactly when we updated to EF core, but that is likely why the other `ReplaceCollectionGroupsAsync` goes out of its way to use the `*Range` methods (it looks like it written in '21). 

I feel updating the `dbContext` collections directly is more readable, so I'd prefer to keep it as is unless you have any objections. _(also fwiw, we use this same pattern in the `UpdateUsersAsync` method in the repository)_

Reply:
Thank you for the reference, all good with me.